crates/re_chunk/src/chunk.rs

```diff
@@ -0,0 +1,478 @@
+                return Err(ChunkError::Malformed {
+                    reason: format!(
+                        "All timelines in a chunk must have the same number of timestamps, matching the number of row IDs.\
+                         Found {} row IDs but {} timestamps for timeline {:?}",
+                        row_ids.len(), time_chunk.times.len(), timeline.name(),
+                    ),
+                });
+            }
+
+            time_chunk.sanity_check()?;
+        }
+
+        // Components
+        for (component_name, list_array) in components {
+            if !matches!(list_array.data_type(), arrow2::datatypes::DataType::List(_)) {
+                return Err(ChunkError::Malformed {
+                    reason: format!(
+                        "The outer array in a chunked component batch must be a sparse list, got {:?}",
+                        list_array.data_type(),
+                    ),
+                });
+            }
+            if let arrow2::datatypes::DataType::List(field) = list_array.data_type() {
+                if !field.is_nullable {
+                    return Err(ChunkError::Malformed {
+                        reason: format!(
+                            "The outer array in chunked component batch must be a sparse list, got {:?}",
+                            list_array.data_type(),
+                        ),
+                    });
+                }
+            }
+            if list_array.len() != row_ids.len() {
+                return Err(ChunkError::Malformed {
+                    reason: format!(
+                        "All component batches in a chunk must have the same number of rows, matching the number of row IDs.\
+                         Found {} row IDs but {} rows for component batch {component_name}",
+                        row_ids.len(), list_array.len(),
+                    ),
+                });
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl ChunkTimeline {
+    /// Returns an error if the Chunk's invariants are not upheld.
+    ///
+    /// Costly checks are only run in debug builds.
+    pub fn sanity_check(&self) -> ChunkResult<()> {
+        let Self {
+            times,
+            is_sorted,
+            time_range,
+        } = self;
+
+        #[allow(clippy::collapsible_if)] // readability
+        if cfg!(debug_assertions) {
+            if *is_sorted != times.windows(2).all(|times| times[0] <= times[1]) {
+                return Err(ChunkError::Malformed {
+                    reason: format!(
+                        "Chunk timeline is marked as {}sorted but isn't: {times:?}",
+                        if *is_sorted { "" } else { "un" },
+                    ),
+                });
+            }
+        }
+
+        #[allow(clippy::collapsible_if)] // readability
+        if cfg!(debug_assertions) {
+            for &time in times {
+                if time < time_range.min() || time > time_range.max() {
```

Comment:
Is time_range allowed to be conservative or should we also be sanity-checking that this is a tight bound?

Reply:
Tighter checks definitely cannot hurt

Reply:
Wait, nevermind I didn't understand your question -- it is a tight bound, isn't it?